### PR TITLE
Fixes #2201 : Fixed checking of declared exceptions.

### LINKED ISSUE #2201: Calling getExceptionTypes() on concrete object that is used as interface doesn't return exception types from interface
org.mockito.internal.stubbing.answers.InvocationInfo#isValidException is doing validation on the exception types that can be thrown from this method.

In my codebase, I'm programming against an interface, but in fact, a spy is created from an instance that is implementing the interface, that drops the exception types that that implementation does not throw. However, the code that is calling the interface is of course unaware of that fact. Also, that implementation is a mock implementation (a real mock implementation, not a Mockito mock()/spy() instance). 

A testcase here:
```
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.exceptions.base.MockitoException;
import org.mockito.junit.jupiter.MockitoExtension;

import java.io.IOException;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.spy;

@ExtendWith(MockitoExtension.class)
public class MockitoExceptionTypesIssue {
    @Test
    void exceptionTypeIoExceptionShouldBeMockableForObjectBeingSpyedOn() {
        final I x = spy(new C());
        final IOException ioException = new IOException("Oh my god!");
        final MockitoException e = assertThrows(
                MockitoException.class,
                () -> doThrow(ioException)
                        .when(x)
                        .m());
        assertEquals("\n" +
                "Checked exception is invalid for this method!\n" +
                "Invalid: " + ioException,
                e.getMessage()
        );
    }

    interface I {
        void m() throws IOException;
    }

    static class C implements I {
        public void m() {
            // no-op
        }
    }
}
```

### PR DESCRIPTION
**Pull request by andrey-kozel**
Fixed checking of declared exceptions. In case when parent contains throws keyword on its method and child overrides this method removing throws, it should be possible to mock throwing exception from child

## Description.
Lets assume that we have the following interface
```
public interface Foo {
        void bar() throws IOException; 
    }
```
And we have class that extends the interface, removing `throws` exception
```
public class Baz implements Foo {
        @Override
        public void bar() {
            
        }
    }
```

When we trying to mock exception throw, it wasn't possible
```
Foo mock = Mockito.spy(new Baz());
Mockito.doThrow(new IOException())
    .when(mock)
    .bar();
```

Now this behavior fixed

## Checklist

 - [x] Read the [contributing guide](https://github.com/mockito/mockito/blob/main/.github/CONTRIBUTING.md)
 - [x] PR should be motivated, i.e. what does it fix, why, and if relevant how
 - [x] If possible / relevant include an example in the description, that could help all readers
       including project members to get a better picture of the change
 - [x] Avoid other runtime dependencies
 - [x] Meaningful commit history ; intention is important please rebase your commit history so that each
       commit is meaningful and help the people that will explore a change in 2 years
 - [x] The pull request follows coding style
 - [x] Mention `Fixes #<issue number>` in the description _if relevant_
 - [x] At least one commit should mention `Fixes #<issue number>` _if relevant_

### CODE REVIEW

**Review by ruslanzianevich — APPROVED** [review]

**Comment by aglumova** [comment]
WOW Amazing!!

File: src/main/java/org/mockito/internal/stubbing/answers/InvocationInfo.java [hunk 1]
```diff
@@ -25,14 +26,33 @@ public InvocationInfo(InvocationOnMock theInvocation) {
     }
 
     public boolean isValidException(Throwable throwable) {
+        if (isValidException(method, throwable)) {
+            return true;
+        }
+
+        return Arrays.stream(method.getDeclaringClass().getInterfaces())
+                .anyMatch(parent -> isValidExceptionForParent(parent, throwable));
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/stubbing/answers/InvocationInfo.java:34]: I am not 100% sure, but I think we should also recursively walk up the parent class hierarchy in case the direct parent doesn't define the method. E.g. I think the current regression tests still fails:

```java
public interface Foo {
    void bar() throws IOException; 
}
public abstract class Baz implements Foo {}
public class SubClass extends Baz {
    @Override
    public void bar() {
    }
}
```

**Review by TimvdLippe — CHANGES_REQUESTED** [review]
Thanks for the PR! I think we are nearly there, but there are some more edge cases we need to handle.

**Comment by andrey-kozel** [comment]
> Thanks for the PR! I think we are nearly there, but there are some more edge cases we need to handle.

Thank you for the comment. I've extended checking for exceptions and added tests for all the cases you've mentioned

**Review by TimvdLippe — APPROVED** [review]
Nice work and regression tests, thanks! 🎉
